presto-main/src/main/java/com/facebook/presto/operator/aggregation/ArrayAggregation.java

```diff
@@ -0,0 +1,202 @@
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, valueType, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type value)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(INPUT_CHANNEL, value), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void input(ArrayAggregationState state, Block value, int position)
+    {
+        if (state.getSliceOutput() == null) {
+            SliceOutput sliceOutput = new DynamicSliceOutput(400);
+            state.setEntries(0);
+            state.setSliceOutput(sliceOutput);
+        }
+        state.setEntries(state.getEntries() + 1);
+        appendTo(state.getType(), state.getSliceOutput(), value, position);
+    }
+
+    private static void appendTo(Type type, SliceOutput output, Block block, int position)
+    {
+        if (type.getJavaType() == long.class) {
+            output.appendLong(type.getLong(block, position));
+        }
+        else if (type.getJavaType() == double.class) {
+            output.appendDouble(type.getDouble(block, position));
+        }
+        else if (type.getJavaType() == Slice.class) {
+            Slice s = type.getSlice(block, position);
+            output.appendInt(s.length());
+            output.appendBytes(s);
+        }
+        else if (type.getJavaType() == boolean.class) {
+            output.appendByte(type.getBoolean(block, position) ? 1 : 0);
+        }
+        else {
+            throw new IllegalArgumentException("Unsupported type: " + type.getJavaType().getSimpleName());
+        }
+    }
+
+    public static void combine(ArrayAggregationState state, ArrayAggregationState otherState)
+    {
+        SliceOutput s1 = state.getSliceOutput();
+        SliceOutput s2 = otherState.getSliceOutput();
+        if (s1 == null && s2 != null) {
+            state.setSliceOutput(s2);
+            state.setEntries(otherState.getEntries());
+        }
+        else if (s1 != null) {
+            s1.appendBytes(s2.slice());
+            state.setEntries(state.getEntries() + otherState.getEntries());
+        }
+    }
+
+    public static void output(ArrayAggregationState state, BlockBuilder out)
+    {
+        if (state.getSliceOutput() == null) {
+            out.appendNull();
+        }
+        else {
+            SliceInput sliceInput = state.getSliceOutput().slice().getInput();
+            Type type = state.getType();
+            long entries = state.getEntries();
+            List<Object> values = toValues(type, sliceInput, entries);
+            Slice s = ArrayType.toStackRepresentation(values);
+            out.writeBytes(s, 0, s.length());
+            out.closeEntry();
+        }
+    }
+
+    private static List<Object> toValues(Type type, SliceInput input, long entries)
+    {
+        List<Object> ret = new ArrayList<Object>((int) entries);
+        for (int i = 0; i < entries; i++) {
```

Comment:
You might want to try rewriting this method as follows, since it could help the VM avoid the if checks on every iteration. You'd need to measure to be sure it has an impact, though -- the VM might be smart enough to hoist the check out of each loop.

```
List<Object> result = new ArrayList<Object>((int) entries);

if (type.equals(...)) { 
    for (int i = 0; i < entries; i++) {
        result.add(input.readLong());
    }
}
else if (type.equals(...)) {
    for (int i = 0; i < entries; i++) {
        ...
    }
}
...
```